File: sdk/AMQP/java/topic/consumer.java
/**
* File: consumer.java
* Description: AMQP protocol. This is consumer code which can get message from exchange and consume them. Topic method.
*
* Author: Stanley
* robomq.io (http://www.robomq.io)
*
*/
import java.io.IOException;
import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.QueueingConsumer;

public class Consumer {
	 private static final String EXCHANGE_NAME = 'exchangeName';

	    public static void main(String[] argv)
	                  throws Exception {

	        ConnectionFactory factory = new ConnectionFactory();
	        factory.setHost('your host');
	        Connection connection = factory.newConnection();
	        Channel channel = connection.createChannel();

	        channel.exchangeDeclare(EXCHANGE_NAME, 'topic');
	        String queueName = channel.queueDeclare().getQueue();

	        String bindingKey = 'routingKey';
	        channel.queueBind(queueName, EXCHANGE_NAME, bindingKey);
	        

	        System.out.println('Waiting for messages. To exit press CTRL+C');

	        QueueingConsumer consumer = new QueueingConsumer(channel);
	        channel.basicConsume(queueName, true, consumer);

	        while (true) {
	            QueueingConsumer.Delivery delivery = consumer.nextDelivery();
	            String message = new String(delivery.getBody());
	            String routingKey = delivery.getEnvelope().getRoutingKey();

	            System.out.println("Received '" + routingKey + '":"' + message + "'");
	        }
	    }

}

File: sdk/AMQP/Ruby/topic/consumer.rb
# File: consumer.rb
# Description: This is the AMQP consumer handles incoming
#     communication from clients publishing messages to a broker server.
#     Messages can be received over AMQP exchange types including one-to-one,
#     from broadcast pattern, or selectively using specified binding key.
#
# Author: Eamin Zhang
# robomq.io (http://www.robomq.io)

require "bunny"

server = "hostname"
port = 5672
vhost = "yourvhost"
username = "username"
password = "password"
exchangeName = "testEx"
queueName = "testQ1"
routingKey = "test.#" #topic with wildcard

while true
  begin
    #connect, disable auto-reconnect so as to manually reconnect
    connection = Bunny.new(:host => server, :port => port, :vhost => vhost, :user => username, :pass => password, :heartbeat => 60, :recover_from_connection_close => false)
    connection.start
    channel = connection.create_channel

    #declare exchange and queue, bind them and consume messages
    exchange = channel.topic(exchangeName, :auto_delete => true)
    queue = channel.queue(queueName, :exclusive => true, :auto_delete => true)
    queue.bind(exchange, :routing_key => routingKey)
    queue.subscribe(:block => false, :manual_ack => false) do |delivery_info, metadata, payload|
      puts payload
    end
    #keep checking the existence of the subscribed queue
    while true
      raise "Lost the subscribed queue %s" % queueName unless connection.queue_exists?(queueName)
      sleep 1
    end
  rescue Exception => e
    #reconnect on exception
    puts "Exception handled, reconnecting...\nDetail:\n%s" % e
    #blindly clean old connection
    begin
      connection.close
    end
    sleep 5
  end
end
